src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/HasChildBookElements.kt

```diff
@@ -0,0 +1,9 @@
+package org.wycliffeassociates.otter.common.data.workbook
+
+import io.reactivex.Observable
+
+interface HasChildBookElements {
```

Comment:
I'm uneasy about the name of this interface (but also ResourceContainerInfo), as it is unclear what it does and what its use is for. 

I roughly know the problem it's intending to solve, but am a little fuzzy just due to how much time has passed- which helps indicate that it doesn't really hold up to being self documenting.

My thought process is mostly as follows: the "has" prefix implies a boolean state: having something or not having something. That something is described by "ChildBookElements," so my expectation is there should be a method I can use that tells me whether or not the object carrying this interface _has_ child book elements. I do not expect the interface to contain such elements, especially if it does not explicitly answer the question the name poses (I suspect the way to check if the object that has this interface actually has child book elements is to do a .isEmpty() call on the list, or to subscribe on children (which now makes my boolean check async (uncomfortable)), but this kinda feels like the interface isn't what is actually providing me the functionality explicitly, despite what the name suggests).

I'm going to tag @natibekele and @KJoslyn here because hopefully me breaking this down is useful for understanding what I value in naming variables/methods/classes/etc and how I interpret them.


Is it intended that this would contain anything other than "helps"? (ie tn, tq?)? If no, the scope can be restricted to just the helps, and doesn't need to be an abstract concept (BookElement).

Reply:
also @mXaln

Reply:
Oh, also see my new tab on draw.io; we might not be on the same page re design.

Reply:
I believe interfaces are typically meant to describe Is-A relationships rather than Has-A, similar to inheritance. I had not seen interfaces with "Has*" before, though it clearly has been done. At the same time, from looking through the github search results I don't really see many large well known projects or repos by companies, even when searching by Java- most are individuals. The Microsoft link is much more compelling, but at the same time I notice that it appears to mostly violate their own style guide: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces

Of course, there is nothing _wrong_ with the code above, it just comes down to code style. I do think that BookElementContainer is much more clear regarding the intent of the interface as well as how it expresses a trait of the class implementing it.

Reply:
Renamed as `BookElementContainer`.